airbyte-integrations/connectors/source-zuora/source_zuora/zuora_client.py

```diff
@@ -0,0 +1,275 @@
+
+            # Casting Zuora ZOQL Data Types into JsonSchema types
+            if self._check_data_type(field_type, ["decimal(22,9)", "integer", "int", "bigint", "timestamp"]):
+                field_type = ["number", "null"]
+            elif self._check_data_type(field_type, ["date", "datetime", "timestamp with time zone", "picklist", "text", "varchar"]):
+                field_type = ["string", "null"]
+            elif self._check_data_type(field_type, ["zoql", "binary", "json", "xml", "blob"]):
+                field_type = ["object", "null"]
+            elif self._check_data_type(field_type, ["boolean"]):
+                field_type = ["boolean", "null"]
+            else:
+                # if there is something else we don't cover, cast it as string
+                field_type = ["string", "null"]
+
+            casted_schema_types.update(**{field_name: {"type": field_type}})
+
+        return casted_schema_types
+
+    # Convert Zuora Fields data types to JSONSchema
+    def _zuora_object_to_json_schema(self, obj: str) -> Dict:
+        self.logger.info(f"Getting schema information for {obj}")
+        raw_data_types = list(self._get_data(q_type="describe", obj=obj))
+        json_schema = []
+        # Get the raw field names and their types
+        for field in range(len(raw_data_types)):
+            json_schema.append({"field": raw_data_types[field]["Column"], "type": raw_data_types[field]["Type"]})
+        # Cast the Zuora Field Types to JsonSchema Types
+        json_schema = self._cast_schema_types(json_schema)
+        return json_schema
+
+    # Convert Zuora Fields data types to JSONSchema
+    def _zuora_list_objects(self) -> List:
+        self.logger.info("Retrieving the list of available Objects from Zuora")
+        raw_data_types = list(self._get_data(q_type="show_tables"))
+        object_list = []
+        # Get the raw field names and their types
+        for field in range(len(raw_data_types)):
+            object_list.append(raw_data_types[field]["Table"])
+        return object_list
+
+    # Method to handle the errors
+    def _handle_requests_exceptions(self, exception: requests.RequestException):
+        exception_code = exception.response.status_code
+        # print(exception_code)
+        if exception_code in [500, 404]:
+            # Bad Request - Server declined the request
+            self.logger.error(f"Bad URL - please check the link, {exception}")
+        elif exception_code == 400:
+            # Bad Request - Server declined the request
+            self.logger.error(f"Bad Request, please check the url query parameters, {exception}")
+        elif exception_code == 401:
+            # Unauthorised or Invalid Token
+            self.logger.error(f"Unauthorised or Expired/Invalid Token, {exception}")
+            # Attempt to refresh the token
+            self.authenticator = ZuoraAuthenticator(self.is_sandbox).generateToken(self.client_id, self.client_secret).get("header")
```

Comment:
Is it possible to check here if the problem with client credentials or just with token expiration ?

Reply:
The problem with client credentials checks on the `check` stage. It covered there.
If we face the issue with an expired or invalid token the ZoqlExportClient will hit the `401 - Unauthorised..` and this is covered in the `_handle_requests_exceptions` method, in particular, to trigger the authentication and generate the new token, then return it back to the ZoqlExportClient and use it from there.